components/esp_http_server/src/httpd_ws.c

```diff
@@ -0,0 +1,311 @@
+{
+    esp_err_t ret = httpd_ws_check_req(req);
+    struct httpd_req_aux *aux = req->aux;
+    if (ret != ESP_OK) {
+        return ret;
+    }
+    if (!pkt) {
+        ESP_LOGW(TAG, LOG_FMT("Argument is invalid"));
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    /* Set the timeout for recv() */
+    struct timeval tv;
+    tv.tv_sec = timeout_sec;
+    tv.tv_usec = 0;
+    if (setsockopt(aux->sd->fd, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv)) < 0) {
+        ESP_LOGW(TAG, LOG_FMT("Failed to set recv() timeout"));
+        return ESP_FAIL;
+    }
+
+    /* Grab the first two bytes */
+    uint8_t init_buf[2] = { 0 };
+    if (httpd_recv_with_opt(req, (char *)init_buf, sizeof(init_buf), false) <= 0) {
+        ESP_LOGW(TAG, LOG_FMT("Failed to receive initial 2 bytes"));
+        return ESP_FAIL;
+    }
+
+    /* Parse the first two bytes */
+    /* Please refer to RFC6455 Section 5.2 for more details */
+    pkt->final = (init_buf[0] & HTTPD_WS_FIN_BIT) != 0;
+    pkt->type = init_buf[0] & HTTPD_WS_OPCODE_BITS;
+    bool masked = (init_buf[1] & HTTPD_WS_MASK_BIT) != 0;
+
+    /* Interpret length */
+    uint8_t init_len = init_buf[1] & HTTPD_WS_LENGTH_BITS;
+    if (init_len < 126) {
+        /* Case 1: If length is 0-125, then this length bit is 7 bits */
+        pkt->len = init_len;
+    } else if (init_len == 126) {
+        /* Case 2: If length byte is 126, then this packet's length bit is 16 bits */
+        uint8_t length_bytes[2] = { 0 };
+        if (httpd_recv_with_opt(req, (char *)length_bytes, sizeof(length_bytes), false) <= 0) {
+            ESP_LOGW(TAG, LOG_FMT("Failed to receive 2 bytes length"));
+            return ESP_FAIL;
+        }
+
+        pkt->len = ((uint32_t)(length_bytes[0] << 8U) | (length_bytes[1]));
+    } else if (init_len == 127) {
+        /* Case 3: If length is byte 126, then this packet's length bit is 64 bits */
+        uint8_t length_bytes[8] = { 0 };
+        if (httpd_recv_with_opt(req, (char *)length_bytes, sizeof(length_bytes), false) <= 0) {
+            ESP_LOGW(TAG, LOG_FMT("Failed to receive 2 bytes length"));
+            return ESP_FAIL;
+        }
+
+        pkt->len = (((uint64_t)length_bytes[0] << 56U) |
+                    ((uint64_t)length_bytes[1] << 48U) |
+                    ((uint64_t)length_bytes[2] << 40U) |
+                    ((uint64_t)length_bytes[3] << 32U) |
+                    ((uint64_t)length_bytes[4] << 24U) |
+                    ((uint64_t)length_bytes[5] << 16U) |
+                    ((uint64_t)length_bytes[6] <<  8U) |
+                    ((uint64_t)length_bytes[7]));
+    }
+
+    /* We only accept SIZE_MAX for now (in ESP-IDF, it's 2^32 - 1) */
+    if (pkt->len > SIZE_MAX) {
+        ESP_LOGW(TAG, LOG_FMT("WS Message too long"));
+        return ESP_ERR_INVALID_SIZE;
+    }
+
+    /* If this packet is masked, dump the mask as well */
+    uint8_t mask_key[4] = { 0 };
+    if (masked) {
```

Comment:
As per the RFC 6455 framing requirements (https://tools.ietf.org/html/rfc6455#section-5.2) all data from client to server should be masked. So, should an error be returned if masked bit is 0?

Reply:
@shahpiyushv hmmm...maybe yes.